### PR TITLE
Add site proofing checks to travis-ci process

### PR DESCRIPTION
**Pull request by SDBrett**
Added site proofing process to travis-ci process.

Added script to detect if files within a directory have changes as part of a commit. Adds capability to determine only run travis jobs relevant to the files changed.

Travis matrix used to enable build tasks to be run in parallel and independently off each other.

Signed-off-by: Brett Johnson <brett@sdbrett.com>

### CODE REVIEW

**Comment by davecheney** [comment]
How critical is this? Jekyll runs inside netlify and they host the result.

**Comment by davecheney** [comment]
Also, I want to explicitly acknowledge your enthusiasm and we’re grateful for the help, but please keep in mind our core contribution principle — talk, then code. Nobody likes to feel their waste their time, and a key to that is to get agreement on making a change before investing that effort. 

Thanks

**Comment by SDBrett** [comment]
> How critical is this? Jekyll runs inside netlify and they host the result.

Low to moderate, the largest area of issue is not to the project itself, but to those contributing to the development of the projects site. Running `bundle install` will install packages as per `Gemfile.lock` and therefore placing potentially vulnerable gems on to those systems.

Good point on the comms, I'll work to raise as issues before running with an idea.

**Comment by davecheney** [comment]
@SDBrett is this PR active? If not, would you please close it.

**Comment by SDBrett** [comment]
It's still active, I'll rebase the branch today

**Comment by SDBrett** [comment]
Performed rebase and PR is ready for review.

File: ciscripts/site-gem-vulnerability-test/cibuild [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+set -e # halt script on error
+
```
Comment (jpeach) [inline ciscripts/site-gem-vulnerability-test/cibuild]: Here, and in the other script, make this independent of the directory it is run from:

```bash
readonly HERE="$(cd $(dirname $0) && pwd)"

cd ${HERE}/../../site # probably need to change this after you move the script
```

File: .travis.yml [hunk 2]
```diff
@@ -1,17 +1,42 @@
...
 
-go:
-  - 1.13.x
+    - language: ruby
+      rvm: 2.6
+      name: Site gem vulnerability check
+      sudo: false # route your build to the container-based infrastructure for a faster build
+      cache: bundler
+      before_script:
+        - chmod +x ./hack/site-gem-vulnerability-test/cibuild
```
Comment (davecheney) [inline .travis.yml]: can you just +x the script in this commit?

Comment (SDBrett) [reply .travis.yml]: Do you mean to change to the below?
```
script:
  - chmod +x ./hack/site-gem-vulnerability-test/cibuild
  - ./hack/site-gem-vulnerability-test/cibuild
```

**Comment by SDBrett** [comment]
> How much time is this going to add to our CI builds?
>
> Can this be moved to Netlify as a post commit step instead?

All three tasks in the travis file run in parallel.

Current process as per master without my changes takes: 3 min 33 sec
Process with my changes takes: 5 min 57 sec

I am looking at options as to running site based checks only if there is a `git diff` within the /site directory. 

As for running with Netlify post commit steps, I do not know. I can have a look at that if it's an option you would like investigated further.

**Comment by davecheney** [comment]
Thank you for your reply. I’m keen to avoid adding 80% to our CI runtime.

**Comment by SDBrett** [comment]
> Thank you for your reply. I’m keen to avoid adding 80% to our CI runtime.

The vulnerability check is very check, under 1 minute. HTML proofer is what adds the most time.

**Comment by davecheney** [comment]
What’s the reason for doing a vulnerabilty check? Doesn’t GitHub do that for us? Also, the site is built by netlify, it’s not a deliverable of the project so I’m not sure what the attack surface is for an vulnerable dependency.

**Comment by SDBrett** [comment]
> What’s the reason for doing a vulnerabilty check? Doesn’t GitHub do that for us? Also, the site is built by netlify, it’s not a deliverable of the project so I’m not sure what the attack surface is for an vulnerable dependency.

The vulnerability check is primarily to assist those working on the site. This prevents them from installing vulnerable gems if they choose to install gems locally. This improves supply chain security which can prevent malicious code injection at build time; though I don't suspect that this has a high probability.


Git not 100% sure on how the scanner works with GitHub, but I know that it did not detect that the Gemfile.lock had a vulnerable version of rubyzip.

**Comment by davecheney** [comment]
I’m finding it difficult  to see the value here — with the exception of myself I thought that most of y’all were using a docker workflow to do your site local builds. 

Even for myself, bundler downloads the gems into a ./vendor directory inside /site. 

What’s the threat model that you are working against?

**Comment by SDBrett** [comment]
> I’m finding it difficult to see the value here — with the exception of myself I thought that most of y’all were using a docker workflow to do your site local builds. Even for myself, bundler downloads the gems into a ./vendor directory inside /site. What’s the threat model that you are working against?

Prevention of vulnerable code on developer systems or being used to build the site.
In your case you would have vulnerable gems within the ./vendor directory if they are specified within Gemfile.lock

**Comment by davecheney** [comment]
But what’s the attack model? The gemfile.lock prevents newer gems — possibly with malware embedded in them— from being downloaded. So that leaves older gems which have vulnerabilities which can be triggered somehow by the Jekyll build process. How can attacker leverage Jekyll to execute malicious code in my system? I’m just not seeing this as something we should do.

**Comment by SDBrett** [comment]
> But what’s the attack model? The gemfile.lock prevents newer gems — possibly with malware embedded in them— from being downloaded. So that leaves older gems which have vulnerabilities which can be triggered somehow by the Jekyll build process. How can attacker leverage Jekyll to execute malicious code in my system? I’m just not seeing this as something we should do.

Gems can be triggered outside the Jekyll build process. A gem on a local system can be executed as part of an attack, it doesn't require the build process. It's also good hygiene. 

From a site specific threat, it is to aid with prevention of malicious code into the website during the build process.

**Comment by davecheney** [comment]
Thank you for working on this. I’m sorry but I’m not convinced this is a likely attack vector. I don’t want to invest ci time and engineering maintenance hours in this process when GitHub already does this albeit at a slightly lower frequency. 

Please remove this part of the pr.

**Comment by davecheney** [comment]
Additionally I’d like to apologise for the back and forth here. It was my fault for not pushing for the discussion to occur on an issue under our contribution guidelines. This would have avoided wasting your time, I’m sorry, I’ll do better in the future.

**Comment by SDBrett** [comment]
> Thank you for working on this. I’m sorry but I’m not convinced this is a likely attack vector. I don’t want to invest ci time and engineering maintenance hours in this process when GitHub already does this albeit at a slightly lower frequency. Please remove this part of the pr.

Will remove from the PR. 

I'll do some more testing on adding the ability to test if files have been changed and put that in it's own issue and PR. The goal being to reduce CI tasks running when not needed, for example only run html-proofer when there are changes in /site/

**Comment by davecheney** [comment]
Thank you. Is there an issue for adding the HTML proofer?

**Comment by SDBrett** [comment]
> Thank you. Is there an issue for adding the HTML proofer?

I couldn't find one, so have created. Issue: #1909

**Comment by davecheney** [comment]
Thank you. Let’s continue the discussion on the issue about how to reduce the ci impact of this test.

File: site/_includes/contributors.html [hunk 3]
```diff
@@ -18,7 +18,7 @@ <h2 class="mb-2">The {{ site.title }} Team</h2>
         <img src="{{ person.image }}" class="rounded-circle" alt="Person" />
         <div class="media-body align-self-center">
           <h6>{{ person.first_name }} {{ person.last_name }}</h6>
-        {{ person.content | markdownify }}    
+        {{ person.content | markdownify }}
```
Comment (jpeach) [inline site/_includes/contributors.html]: Can you please revert this change?

File: hack/travis/travis-check-files-changed.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
+# Checks if the files changed in the last commit are contained within specifed directories.
```
Comment (jpeach) [inline hack/travis/travis-check-files-changed.sh]: This should have a `#!/usr/bin/env bash` too.

File: .travis.yml [hunk 5]
```diff
@@ -1,17 +1,36 @@
...
+      cache: bundler
+      env:
+        - NOKOGIRI_USE_SYSTEM_LIBRARIES=true # speeds up installation of html-proofer
+        - SEARCH_DIRECTORIES="site"
+      addons:
+        apt:
+          packages:
+            - libcurl4-openssl-dev
+      before_install:
+        - chmod +x ./hack/travis/travis-check-files-changed.sh
```
Comment (jpeach) [inline .travis.yml]: You should not need to chmod these. Git preserves permissions, so chmod the script and commit them and they will still be executable.

File: hack/travis/travis-check-files-changed.sh [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+#      install:
+#        - ./hack/travis/travis-check-files-changed.sh $SEARCH_DIRECTORIES ; RETURN_CODE=$? ; if [ $RETURN_CODE -eq 137 ]; then travis_terminate 0; elif [ $RETURN_CODE -ne 0 ]; then travis_terminate $RETURN_CODE; fi
+
+
+# 1. Get all the arguments of the script
+# https://unix.stackexchange.com/a/197794
+PATHS_TO_SEARCH="$*"
+
+# 2. Make sure the paths to search are not empty
+if [ -z "$PATHS_TO_SEARCH" ]; then
```
Comment (jpeach) [inline hack/travis/travis-check-files-changed.sh]: `$#` gives you the number of arguments to the script. 
```bash
if [ $# -eq 0 ]; then
    usage
    exit 1
fi
```

File: hack/travis/travis-check-files-changed.sh [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    echo "Please provide the paths to search for."
+    echo "Example usage:"
+    echo "./exit-if-path-not-changed.sh path/to/dir1 path/to/dir2"
```
Comment (jpeach) [inline hack/travis/travis-check-files-changed.sh]: This isn't the name of the script any more :)

File: hack/travis/travis-check-files-changed.sh [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    exit 1
+fi
+
+# 3. Get the latest commit
+LATEST_COMMIT=$(git rev-parse HEAD)
+
+# 4. Get the latest commit in the searched paths
+LATEST_COMMIT_IN_PATH=$(git log -1 --format=format:%H --full-diff $PATHS_TO_SEARCH)
```
Comment (jpeach) [inline hack/travis/travis-check-files-changed.sh]: I think what you want here is:
```bash
LATEST_COMMIT_IN_PATH=$(git log -1 --format=format:%H --full-diff "$@")
```

https://www.gnu.org/software/bash/manual/html_node/Special-Parameters.html

File: hack/travis/travis-check-files-changed.sh [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+#        - SEARCH_DIRECTORIES="site"
+#      before_install:
+#        - chmod +x ./hack/travis/travis-check-files-changed.sh
+#      install:
+#        - ./hack/travis/travis-check-files-changed.sh $SEARCH_DIRECTORIES ; RETURN_CODE=$? ; if [ $RETURN_CODE -eq 137 ]; then travis_terminate 0; elif [ $RETURN_CODE -ne 0 ]; then travis_terminate $RETURN_CODE; fi
+
+
+# 1. Get all the arguments of the script
+# https://unix.stackexchange.com/a/197794
+PATHS_TO_SEARCH="$*"
```
Comment (jpeach) [inline hack/travis/travis-check-files-changed.sh]: Can noted below, I think you want to remove this and just use `"$@"`.

File: hack/travis/travis-check-files-changed.sh (hunk 4 above)
Comment (jpeach) [reply hack/travis/travis-check-files-changed.sh]: and a `set -e # halt script on error`

**Comment by jpeach** [comment]
@davecheney FYI, I'm planning to work with @SDBrett to land this today. If you have concerns, ping me.

**Comment by davecheney** [comment]
I would prefer we don't get into this business. Github covers this sufficiently.

**Comment by jpeach** [comment]
> On Nov 29, 2019, at 8:37 AM, Dave Cheney <notifications@github.com> wrote:
> 
> ﻿I would prefer we don't get into this business. Github covers this sufficiently.

Does GitHub cover link checking with HTML proofer? Where’s that set up?

**Comment by davecheney** [comment]
I’m sorry I think I misentetpreted this change from the title. If the gen scanning stuff is removed, great. 

What is the cost to ci time from link checking? Do we need to run it on every commit? Can it be done on the netlify side? Can it be done in batch?

**Comment by jpeach** [comment]
> On 29 Nov 2019, at 9:09 am, Dave Cheney <notifications@github.com> wrote:
> 
> I’m sorry I think I misentetpreted this change from the title. If the gen scanning stuff is removed, great. 
> 
> What is the cost to ci time from link checking? Do we need to run it on every commit? Can it be done on the netlify side? Can it be done in batch?

* The time cost is approximately that of `make site-check`.
* The "travis-check-files-changed.sh" script is trying to ensure that this check only runs when website files change, so the cost to the majority of PRs should be zero.
* I don't know whether netlify does this or not.
* Not sure what you mean by doing it in batch?

**Comment by davecheney** [comment]
On Fri, 29 Nov 2019 at 09:49, James Peach <notifications@github.com> wrote:

>
>
> > On 29 Nov 2019, at 9:09 am, Dave Cheney <notifications@github.com>
> wrote:
> >
> > I’m sorry I think I misentetpreted this change from the title. If the
> gen scanning stuff is removed, great.
> >
> > What is the cost to ci time from link checking? Do we need to run it on
> every commit? Can it be done on the netlify side? Can it be done in batch?
>
> * The time cost is approximately that of `make site-check`.
> * The "travis-check-files-changed.sh" script is trying to ensure that this
> check only runs when website files change, so the cost to the majority of
> PRs should be zero.
> * I don't know whether netlify does this or not.
> * Not sure what you mean by doing it in batch?


As in, doing this outside the CI loop.
